warehouse/query-core/src/main/java/datawave/query/transformer/UniqueTransform.java

```diff
@@ -303,4 +355,207 @@ public void funnel(byte[] from, PrimitiveSink into) {
             into.putBytes(from);
         }
     }
+
+    /**
+     * An iterator of documents for this unique transform given an underlying iterator of documents.
+     */
+    public class UniqueTransformIterator implements Iterator<Map.Entry<Key,Document>> {
+        private final Iterator<Map.Entry<Key,Document>> iterator;
+        private Map.Entry<Key,Document> next = null;
+
+        public UniqueTransformIterator(Iterator<Map.Entry<Key,Document>> iterator) {
+            this.iterator = iterator;
+        }
+
+        @Override
+        public boolean hasNext() {
+            if (next == null) {
+                next = getNext();
+            }
+            return (next != null);
+        }
+
+        @Override
+        public Map.Entry<Key,Document> next() {
+            Map.Entry<Key,Document> o = null;
+            if (next == null) {
+                o = getNext();
+            } else {
+                o = next;
+                next = null;
+            }
+            return o;
+        }
+
+        private Map.Entry<Key,Document> getNext() {
+            Map.Entry<Key,Document> o = null;
+            while (o == null && iterator.hasNext()) {
+                o = apply(iterator.next());
+            }
+            // see if there are any results cached by the transform
+            if (o == null) {
+                o = flush();
+            }
+            return o;
+        }
+
+    }
+
+    /**
+     * A builder of unique transforms
+     */
+    public static class Builder {
+        private UniqueFields uniqueFields;
+        private Comparator<byte[]> keyComparator;
+        private FileSortedMap.RewriteStrategy<byte[],Document> keyValueComparator;
+        private QueryModel model;
+        private int bufferPersistThreshold;
+        private List<IvaratorCacheDirConfig> ivaratorCacheDirConfigs;
+        private String hdfsSiteConfigURLs;
+        private String subDirectory;
+        private int maxOpenFiles;
+        private int numRetries;
+        private long queryExecutionForPageTimeout;
+        private FileSortedSet.PersistOptions persistOptions;
+
+        public Builder() {
+            keyComparator = new ByteArrayComparator();
+
+            keyValueComparator = (key, original, update) -> {
+                long ts1 = getTimestamp(original);
+                long ts2 = getTimestamp(update);
+                return (ts2 > ts1);
+            };
+        }
+
+        /**
+         * Build a list of potential hdfs directories based on each ivarator cache dir configs.
+         *
+         * @param ivaratorCacheDirConfigs
+         * @param hdfsSiteConfigURLs
+         * @param subdirectory
+         * @return A path
+         * @throws IOException
+         *             for issues with read/write
+         */
+        private static List<IvaratorCacheDir> getIvaratorCacheDirs(List<IvaratorCacheDirConfig> ivaratorCacheDirConfigs, String hdfsSiteConfigURLs,
```

Comment:
this looks like a copy/paste from IteratorBuildingVisitor. Can you shift to a common function?

Reply:
They are very similar, but different enough that I think they warrent separate implementations.